assignments/tdecacqu/notes-12.org

```diff
@@ -0,0 +1,105 @@
+#+TITLE: SICP 2.4 assignment
+#+AUTHOR: Tristan de Cacqueray
+#+BABEL: :cache yes
+#+PROPERTY: header-args :tangle yes
+
+* Building Abstractions with Data
+** Data Representation
+*** Exercise 2.73
+1/ The deriv procedure has been refactored to use the apply-operator technique.
+It couldn't be used for number? and variable? because those are primitive type
+and there is no operation to dispatch.
+
+2/ The procedures for sums and products are:
+
+#+BEGIN_SRC scheme
+  (define (deriv-sum exp var)
+    (make-sum (deriv (addend exp) var) (deriv (augend exp) var)))
+  (put 'deriv '+ deriv-sum)
+
+  (define (deriv-product exp var)
+    (make-sum
+     (make-product
+      (multiplier exp)
+      (deriv (multiplicand exp) var))
+     (make-product
+      (deriv (multiplier exp) var)
+      (multiplicand exp))))
+  (put 'deriv '* deriv-product)
+#+END_SRC
+
+3/ Exponentiation can be installed like so:
+
+#+BEGIN_SRC scheme
+  (put 'deriv '**
+       (lambda (exp var)
+         (make-product
+          (make-product (exponent exp)
+                        (make-exponentiation (base exp)
```

Comment:
Wouldn't these actually fail? There's no definition of ``make-product`` or ``make-exponentiation``. Shouldn't it be fetched like ``(define make-exponentiation (get 'deriv '** 'make-exponentiation)))`` or equivalent?

Reply:
Although I guess that depends how far we want to take this example..

Reply:
I'm not sure we actually need to install 'make-product or 'make-exponentiation type dispatch because it's only used locally (e.g. you wouldn't need a '** 'make-product). I assumed this code was evaluated after the program in 2.3.2.

Though I haven't tested that exercise.